A27-xds-global-load-balancing.md

```diff
@@ -0,0 +1,440 @@
+XdsClient object.
+
+#### CDS LB Policy
+
+The CDS LB policy will be the top-level LB policy.  Its configuration
+will be of the following form:
+
+```
+{
+  "cluster": "<cluster name>"
+}
+```
+
+The CDS policy will use the XdsClient object passed in from the `xds` resolver
+to query for the `Cluster` resource with the cluster name from the LB
+policy configuration.
+
+Initially, the CDS policy will always create an EDS policy as the child
+policy.  In the future, this may change; for example, we may add support
+for clusters that use DNS or for aggregate clusters, which would require
+different child policies.
+
+#### EDS LB Policy
+
+The EDS LB policy will be the child of the CDS LB policy.  Its
+configuration will be of the following form:
+
+```
+{
+  "edsServiceName": "<EDS service name>",
+  // Optional; if not specified, load reporting will be disabled.
+  // If value is the empty string, that means to use the xDS server as
+  // the LRS server.  (Currently, we do not support non-empty values;
+  // support for this will be added later.)
+  "lrsLoadReportingServerName": "<LRS server name>"
+}
+```
+
+The EDS policy will use the XdsClient object passed in from the xds
+resolver to query for the `ClusterLoadAssignment` resource with the
+EDS service name from the LB policy configuration.
+
+Note that unlike the old grpclb policy, which just received a flat list
+of backend servers from the balancer to round-robin across, the EDS policy 
+receives a hierarchical list, where the servers are grouped into
+localities, and the localities are grouped into priorities.  The EDS
+policy will be responsible for both selecting the highest priority set
+of localities that are reachable and distributing the traffic across the
+localities in the selected priority based on the locality weights.
+
+Note that the EDS policy will *not* support
+[overprovisioning](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/overprovisioning),
+which is different from Envoy.  Envoy takes the
+overprovisioning into account in both [locality-weighted load
+balancing](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/locality_weight)
+and [priority
+fail-over](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/priority),
+but gRPC assumes that the xDS server will update it to redirect traffic
+when this kind of graceful fail-over is needed.
+
+Unlike the grpclb policy, which relied on server-side load reporting to
+the balancer, the EDS policy will perform client-side load reporting
+using the [LRS
+protocol](https://github.com/envoyproxy/data-plane-api/blob/master/envoy/service/load_stats/v2/lrs.proto).
+Note that the EDS policy will not support per-endpoint stats; it will
+report only per-locality stats.  Currently, the EDS policy will report
+only client-side stats; in the future, we will support also reporting
+backend server stats reported to the client via
```

Comment:
Will the backend server stats be passed to the client use in-band way here?

Reply:
We will eventually support sending backend stats to the client both on a per-RPC basis and via an out-of-band stream.  The per-RPC mechanism, which involves the backend load data being sent in each RPC's trailing metadata, will probably happen first.  However, the OOB mechanism is useful for cases where the RPCs are long-running streams, where we need the data to be reported more frequently than RPCs are completing.

In any case, the details for backend load reporting will be covered in another gRFC when we get to the point where we're ready to implement this.

Reply:
Thanks, your answer helped me a lot.